src/session/mpi.rs

```
@@ -152,31 +162,45 @@ impl SessionMPI {
         self.providers.first().expect("no providers")
     }
 
-    pub fn hostfile(&self) -> String {
-        let peers = &self.providers;
+    /// Creates a hostfile for all connected providers
+    pub fn hostfile(&self, threads_per_proc: usize) -> String {
+        let mut peers: Vec<&ProviderMPI> = self.providers.iter().collect();
+        // The sort will become unnecessary once golem-unlimited gets
+        // some scheduling capabilities.
+        peers.sort_unstable_by_key(|peer| peer.num_cores());
         let file_lines: Vec<_> = peers
             .iter()
             .map(|peer| {
-                let ip_sock = &peer.info.peer_addr;
-                let ip_sock: SocketAddr = ip_sock
-                    .parse()
-                    .unwrap_or_else(|_| panic!("GU returned an invalid IP address, {}", ip_sock));
-                let ip = ip_sock.ip();
+                let ip = {
+                    let ip_sock = &peer.info.peer_addr;
+                    let ip_sock: SocketAddr = ip_sock.parse().unwrap_or_else(|_| {
+                        panic!("GU returned an invalid IP address, {}", ip_sock)
```

Do you we really want to panic here or could we throw an error perhaps?